ginivision/src/main/java/net/gini/android/vision/internal/cache/MemoryCache.java

```diff
@@ -19,64 +26,141 @@
 
 public abstract class MemoryCache<K, V> {
 
+    private static final boolean DEBUG = false;
+    private final Logger mLog;
     private final LruCache<K, V> mCache;
     private final Queue<Worker<K, V>> mWorkerQueue = new LinkedList<>();
     private final List<Worker<K, V>> mRunningWorkers;
     private final int mRunningWorkersLimit;
+    private final Map<K, List<AsyncCallback<V>>> mWaitingCallbacks = new HashMap<>();
 
     public MemoryCache(final int runningWorkersLimit) {
         mRunningWorkersLimit = runningWorkersLimit;
         mRunningWorkers = new ArrayList<>(runningWorkersLimit);
         mCache = createCache();
+        if (DEBUG) {
+            mLog = LoggerFactory.getLogger(getClass());
+        } else {
+            mLog = NOPLogger.NOP_LOGGER;
+        }
     }
 
     protected abstract LruCache<K, V> createCache();
 
-    public void get(@NonNull final Context context, @NonNull final K key, @NonNull final
-    AsyncCallback<V> callback) {
+    public void get(@NonNull final Context context, @NonNull final K key,
+            @NonNull final AsyncCallback<V> callback) {
+        mLog.debug("Get for key {}", getNameForLog(key));
         if (mCache.get(key) != null) {
-            callback.onSuccess(mCache.get(key));
+            final V value = mCache.get(key);
+            mLog.debug("Return cached {}", getNameForLog(value));
+            callback.onSuccess(value);
             return;
         }
 
-        final Worker worker = createWorker(mRunningWorkers, key,
+        mLog.debug("Create worker");
+        final Worker<K, V> worker = createWorker(mRunningWorkers, key,
                 new AsyncCallback<V>() {
                     @Override
                     public void onSuccess(final V result) {
+                        mLog.debug("Worker finished with result {}", getNameForLog(result));
                         mCache.put(key, result);
-                        callback.onSuccess(result);
+                        final List<AsyncCallback<V>> callbacks = mWaitingCallbacks.get(key);
```

Comment:
Both the `onSuccess` and `onError` methods has almost the code, except for the invocation to the callback. Could you get rid of this duplicated code?